src/react/hooks/useLatestPropsOnEffect.ts

```diff
@@ -0,0 +1,31 @@
+import type { BoardProps } from 'boardgame.io/react';
+import { useEffect, useState } from 'react';
+import { useEffectListener } from './useEffectListener';
+import { useBoardProps } from './useBoardProps';
+
+/**
+ * Returns the latest board props when one or more effect
+ * is triggered. Essentially sugar around `useEffectListener`.
+ * @param effectTypes - List of effects to subscribe to.
+ * @return The boardgame.io props including G and ctx
+ */
+export function useLatestPropsOnEffect(...effectTypes: Array<string>) {
+  const boardProps = useBoardProps();
+  const [props, setProps] = useState(boardProps);
+
+  useEffectListener(
+    '*',
+    (effectName: string, _payload: any, boardProps: BoardProps) => {
+      if (effectTypes.includes(effectName) || effectTypes.includes('*')) {
+        setProps(boardProps);
+      }
+    },
+    [effectTypes]
```

Comment:
I think as you suggested, the spread array will have a different identity each run, so we need to memoize based on the individual arguments. Shouldn’t that be as simple as doing this?

```suggestion
    [...effectTypes]
```

Reply:
So I did think of that solution but I believe the "rules of hook" linting would complain. In practice, I think the spread solves the issue quite elegantly but maybe it needs an additional comment to explain why since it's not the standard practice?

Reply:
Sure, or we could use a `useMemo` call on the array, but I would guess that the same issue might crop up? Is providing `effectTypes` directly as the dependency array permitted instead?

Reply:
So I tried to read about this a bit but haven't found loads... [This]( https://stackoverflow.com/questions/59107782/react-useeffect-has-a-spread-element-in-its-dependency-array) suggests using the spread as the deps array but people seem to warn it's not "correct". 

Another way to do this might be to use `useRef` and check equality:

```
const effects = useRef(effectTypes)

if (
	effectTypes.length !== effects.current.length || 
	!effects.current.every((v, i) => v === effectTypes[i])
) {
	effects.current = effectTypes
}
```

If we're referencing `effects.current.includes(name)` in the callback and put `[effects]` as the dependency array I believe the latest value will be read so that should be fine. 

Now, I've just trialled all 3 solutions and they all work. Given the React doc has [an example](https://reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) (not recommended but edge case) featuring the spread in the array I guess we can go with that as it's still what seems the most reasonable solution. If we want to be "correct" we can use the `useRef` method.

Reply:
Ah, you just beat me to it!

I also tried to do some more digging, and didn’t find much, but [this comment][c] is interesting:

> …it's not supported to have deps of different lengths between renders…

[c]: https://github.com/facebook/react/issues/15884#issuecomment-522624757

In other words, this usage would apparently not be OK regardless of how we passed the arguments:

```js
const effects = flag ? ['foo'] : ['bar', 'baz'];
useLatestPropsOnEffect(...effects);
```

This is a bit of a head scratcher but I also arrived at something like you suggested using state, but a ref is even better to avoid pointless rerenders 👍